### PR TITLE
Switch to using poetry

### PR DESCRIPTION
**Pull request by devonh**
This brings sygnal in line with synapse & allows the use of a poetry.lock file to track dependency versions more precisely.

### CODE REVIEW

**Review by erikjohnston — CHANGES_REQUESTED** [review]
Thank you! This looks good, but I think we need to update the dockerfile to use poetry?

**Comment by devonh** [comment]
> Thank you! This looks good, but I think we need to update the dockerfile to use poetry?

Ah! I knew I must be missing something. Will do.

File: docker/Dockerfile [hunk 1]
```diff
@@ -6,26 +6,81 @@
...
+###
+### Stage 0: generate requirements.txt
+###
+# We hardcode the use of Debian bookworm here because this could change upstream.
+FROM docker.io/library/python:${PYTHON_VERSION}-slim-bookworm as requirements
+
+# We install poetry in its own build stage to avoid its dependencies conflicting with
+# sygnal's dependencies.
+RUN --mount=type=cache,target=/root/.cache/pip \
+  pip install --user "poetry==1.3.2"
```
Comment (erikjohnston) [inline docker/Dockerfile:20]: That's quite an old version of poetry to pin to? It's now at 1.8.3

**Review by erikjohnston — COMMENTED** [review]
Otherwise looks good. I assume you've manually tested building and running the docker image?

File: docker/Dockerfile (hunk 1 above)
Comment (devonh) [reply docker/Dockerfile:20]: I just grabbed the same version from synapse. I'll bump it up here since I don't think we have any of the same constraints.

**Pull request merged**